### PR TITLE
publishing multi-arch image

### PR DESCRIPTION
**Pull request by morlay**
example:  https://hub.docker.com/r/morlay/verdaccio/tags?page=1&ordering=last_updated
```
docker buildx build --push --platform=linux/amd64,linux/arm64 --tag morlay/verdaccio .
```

github actions setup 
https://github.com/docker/setup-buildx-action

I haven't found the publish workflow.

### CODE REVIEW

**Comment by juanpicado** [comment]
hi @morlay thanks for your PR. I'd need you to share some info regarding what is this about, I recall this topic was mentioned months ago, but I ignore the trade-offs of merging this. I tagged collaborators and let's wait.

**Comment by morlay** [comment]
@juanpicado just want to publish verdaccio as multi arch docker image

so people could use verdaccio image in arm64 device like Apple M1

**Comment by morlay** [comment]
switch to draft until whole workflow ready

**Comment by morlay** [comment]
@juanpicado 
could you show me where the final workflow for publishing docker image.

i haven't found it 😂

**Comment by juanpicado** [comment]
It happens directly in docker hub. We don't have this flow in GitHub actions yet.

**Comment by morlay** [comment]
could i add someone by https://github.com/docker/setup-buildx-action ？

**Comment by juanpicado** [comment]
I don't know yet, but what I think we cannot do is modify the Dockerifle, correct me if I'm wrong, that might break the `latest` tag. The only I know about this topic is https://www.docker.com/blog/apple-silicon-m1-chips-and-docker/ and seems pretty new.  So let's wait. We are in the holiday season 🙃  so I don't expect much movement on this in the next weeks.

**Comment by morlay** [comment]
@juanpicado have a nice holiday first. I just put this PR here.

Apple M1 is one case. in fact, in cloud compute, arm marchine has raised.
some testing https://recap.dev/blog/arm-backed-servers-better-performance-for-less-money/ 

I added a publish workflow.

workflow for master branch https://github.com/morlay/verdaccio/runs/1561104741 (now as `edge` tag, could rename by `tag-custom` and `tag-custom-only` https://github.com/crazy-max/ghaction-docker-meta#inputs)
workflow for tag https://github.com/morlay/verdaccio/runs/1561107479

https://hub.docker.com/r/morlay/verdaccio/tags?page=1&ordering=last_updated

btw is image tag `4.x-next` for master branch ?

**Comment by juanpicado** [comment]
> btw is image tag 4.x-next for master branch ?

Yes, exactly. 

Furthermore, during the testing phase, I publish `5.x` to GitHub packages 

https://github.com/verdaccio/verdaccio/packages/13531

but this one is not relevant yet and not recommendable to use it.

**Comment by morlay** [comment]
@juanpicado got it. 

one more thing, `docker.pkg.github.com` is deprecated & not support publishing the multi-arch image.
https://docs.github.com/en/free-pro-team@latest/packages/guides/migrating-to-github-container-registry-for-docker-images

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+          registry: docker.io
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Prepare docker image tags
+        id: docker_meta
+        uses: crazy-max/ghaction-docker-meta@v1
+        with:
+          images: ${{ github.repository }}
+          tag-custom: 4.x-next
+          tag-custom-only: ${{ github.ref == 'refs/heads/master' }}
```
Comment (morlay) [inline .github/workflows/docker-publish.yml:37]: master branch should publish tag `4.x-next`

File: .github/workflows/docker-publish.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          tag-semver: |
+            {{version}}
+            {{major}}
+            {{major}}.{{minor}}
+      - name: Build & Push
```
Comment (morlay) [inline .github/workflows/docker-publish.yml:42]: tag rule https://github.com/crazy-max/ghaction-docker-meta#handle-semver-tag

when tag `v4.10.1`
the tag `latest` `4` `4.10`  `4.10.1` will published

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM node:14.15.1-alpine as builder
+FROM --platform=${BUILDPLATFORM:-linux/amd64} node:14.15.1-alpine as builder
```
Comment (morlay) [inline Dockerfile:1]: limit build only run once on same arch as host

only need final image as multi-arch image.

File: .github/workflows/docker-publish.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      - '**'
+    tags:
+      - 'v*'
+
+jobs:
+  docker:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: docker/setup-qemu-action@v1
```
Comment (morlay) [inline .github/workflows/docker-publish.yml:21]: Dockerfile contains `RUN`, we need this run execute `arm64` binaries on amd64 host.

**Review by morlay — COMMENTED** [review]
@juanpicado 

only need add secrets `DOCKER_USERNAME`  and `DOCKER_PASSWORD`.
We could switch to github actions for multi-arch image publishing with same tag rule.

**Comment by juanpicado** [comment]
> @juanpicado
> 
> only need add secrets `DOCKER_USERNAME` and `DOCKER_PASSWORD`.
> We could switch to github actions for multi-arch image publishing with same tag rule.

ok, this looks much better, let me get into it.

File: .github/workflows/docker-publish.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+name: Docker publish to docker.io
+
+on:
+  push:
+    paths:
+      - .github/workflows/docker-publish.yml
+      - 'packages/**'
+      - 'docker-bin/**'
+      - 'package.json'
+      - 'pnpm-*.yaml'
```
Comment (juanpicado) [inline .github/workflows/docker-publish.yml:10]: master branch uses yarn, in this case would be `yarn.lock`. Remember the master branch structure is completely difference, this looks like you took it from the branch `5.x`. At this point we want only cover `master === 4.x`.

**Review by juanpicado — COMMENTED** [review]
After some testing it works pretty fine :+1: 
docker1

File: .github/workflows/docker-publish.yml (hunk 5 above)
Comment (juanpicado) [reply .github/workflows/docker-publish.yml:10]:
```
  push:
    paths:
      - .github/workflows/docker-publish.yml
      - 'src/**'
      - 'conf/**'
      - 'types/**',
      - 'docker-bin/**'
      - 'bin/**' 
      - 'package.json'
      - 'yarn.lock'
```

**Review by juanpicado — CHANGES_REQUESTED** [review]
Small change and we can move on this :+1:

**Comment by morlay** [comment]
@juanpicado updated with your suggestion.

btw. after merging, need to disable build of Docker Hub too.

**Review by juanpicado — APPROVED** [review]
Great work @morlay !

**Comment by juanpicado** [comment]
@morlay is there anything worth mention at the documentation? anything aside of the new platform supported the users must know from now on? 

https://verdaccio.org/docs/en/docker

**Comment by juanpicado** [comment]
There it is :) :1st_place_medal: 

Screenshot from 2020-12-29 11-46-07

**Comment by morlay** [comment]
> @morlay is there anything worth mention at the documentation? anything aside of the new platform supported the users must know from now on?
> 
> https://verdaccio.org/docs/en/docker

Nope. nothing need to changes. 

Multi-arch image return manifests list with platforms.
```
$ docker manifest inspect verdaccio/verdaccio:4.x-next
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2410,
         "digest": "sha256:03ae6df781a2ea6ce733bcb7fbc5b37e1ed47d101bc7ac768ce8873e167a5053",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2410,
         "digest": "sha256:af4fc662cdbf232cf14815d53d6de58e5daa604ad79d45406f1d465017c50195",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```

when we run `docker pull`, it will return the matched digest

```bash
$ docker pull verdaccio/verdaccio:4.x-next

# on amd64 device, will be
docker pull verdaccio/verdaccio@sha256:03ae6df781a2ea6ce733bcb7fbc5b37e1ed47d101bc7ac768ce8873e167a5053

# on arm64 device, will be
docker pull verdaccio/verdaccio@sha256:af4fc662cdbf232cf14815d53d6de58e5daa604ad79d45406f1d465017c50195
```

**Pull request merged**